llvm/labs/lab2/sharapov_georgiy/LoopFramer.cpp

```diff
@@ -0,0 +1,72 @@
+#include "llvm/IR/Dominators.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+
+struct LoopFramer : public llvm::PassInfoMixin<LoopFramer> {
+  llvm::PreservedAnalyses run(llvm::Function &F,
+                              llvm::FunctionAnalysisManager &FAM) {
+    llvm::LLVMContext &Context = F.getContext();
+    llvm::Module *ParentModule = F.getParent();
+
+    llvm::FunctionType *funcType =
+        llvm::FunctionType::get(llvm::Type::getVoidTy(Context), false);
+    llvm::FunctionCallee loopStart =
+        ParentModule->getOrInsertFunction("loop_start", funcType);
+    llvm::FunctionCallee loopEnd =
+        ParentModule->getOrInsertFunction("loop_end", funcType);
+
+    llvm::LoopAnalysis::Result &LI = FAM.getResult<llvm::LoopAnalysis>(F);
+    for (auto *L : LI) {
+      insertLoopStartEnd(L, loopStart, loopEnd);
+    }
+
+    auto PA = llvm::PreservedAnalyses::all();
+    PA.abandon<llvm::LoopAnalysis>();
+    PA.abandon<llvm::DominatorTreeAnalysis>();
```

Comment:
Why do you think it should be abandoned too?

Reply:
I thought that if we add calls to some functions, then we are changing the control flow and, consequently, we can affect dominance of blocks.

Reply:
Dominance can be changed if we add/move/delete blocks. Calls don't change blocks in caller function.